Initial implementation.

sjrd opened the pull request:
This plugin can be used as follows (or will be, once published):
https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/compare/getting-started-tutorial-series...vite-plugin

This is as a follow-up to https://github.com/scala-js/scala-js-website/pull/590#discussion_r1096551638

---

Regarding tests: they directly interact with the plugin, without going through Vite or Rollup. This seems to be standard practice among all the Vite and/or Rollup plugins I looked at. Apparently, integration tests (like with have with sbt `scripted`) are not a thing in that universe.

sjrd commented:
Green build here: https://github.com/sjrd/vite-plugin-scalajs/actions/runs/4306121401/jobs/7509511713

sjrd commented:
@gzm0 Do you think you might have some time for this PR in the coming week? Should I ask someone else to review?

Does this plugin make sense to you overall, in terms of what it offers?

gzm0 commented:
Sorry, I missed this over the weekend. Given the current forecast for the weekend, I'm quite confident I can find the time to review this. Is that enough?

sjrd commented:
> Given the current forecast for the weekend

:smile: 

Yes, sure, that's enough. Thank you!

sjrd commented:
As an additional confidence test, here is a "real-world" usage of the plugin: https://github.com/scalacenter/scastie/pull/790

gzm0 commented on tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+  },
+  "include": [
+    "./*.ts"
+  ],
```
Consider leaving the default (`include: ["**/*"]`).

gzm0 commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+  });
+
+  return new Promise((resolve, reject) => {
+    child.on('close', code => {
```
Does this event fire if the process cannot be spawned at all?

sjrd replied:
No! Good catch. I added an `'error'` handler for that.

gzm0 commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+    cwd: cwd,
+  };
+  const child = process.platform === 'win32'
+    ? spawn("sbt.bat", args.map(x => `"${x}"`), {shell: true, ...options})
```
Have you considered the following options for win:

> windowsVerbatimArguments [<boolean>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) No quoting or escaping of arguments is done on Windows. Ignored on Unix. This is **set to true automatically when shell is specified** and is CMD. Default: false.
>
> windowsHide [<boolean>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) Hide the subprocess console window that would normally be created on Windows systems. Default: false.

sjrd replied:
I tried many different things on Windows.

`windowsVerbatimArguments` did not seem to make a difference.

I did not try `windowsHide` but the default behavior seems to do what I want, because we're already running in a console, so it executes in the same console window.

gzm0 commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+function printSbtTask(task: string, cwd?: string): Promise<string> {
+  const args = ["--batch", "-no-colors", "-Dsbt.supershell=false", `print ${task}`];
+  const options: SpawnOptionsWithoutStdio = {
+    cwd: cwd,
```
consider: `stdio: { 'ignore', 'pipe', 'inherit' }`. You can then drop the `on` handler below for stderr.

gzm0 commented on README.md:
```diff
@@ -2,4 +2,114 @@
+
+## Importing `@JSExportTopLevel` Scala.js members
+
+`@JSExportTopLevel("foo")` members in the Scala.js code are exported from the modules that Scala.js generate.
```
```suggestion
`@JSExportTopLevel("foo")` members in the Scala.js code are exported from the modules that Scala.js generates.
```

gzm0 commented on README.md:
```diff
@@ -2,4 +2,114 @@
+import scala.scalajs.js
+import scala.scalajs.js.annotation._
+
+@JSExportTopLevel("square")
```
```suggestion
@JSExportTopLevel("ScalaJSLib")
```

gzm0 commented on README.md:
```diff
@@ -2,4 +2,114 @@
+### Exports in other modules
+
+By default, `@JSExportTopLevel("Foo")` exports `Foo` from the `main` module, which is why we import from `scalajs:main.js`.
+We can also split the Scala.js exports in several modules.
```
```suggestion
We can also split the Scala.js exports into several modules.
```
?

gzm0 commented on README.md:
```diff
@@ -2,4 +2,114 @@
+import scala.scalajs.js
+import scala.scalajs.js.annotation._
+
+@JSExportTopLevel("square", "library")
```
```suggestion
@JSExportTopLevel("ScalaJSLib", "library")
```

gzm0 commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,47 @@
+name: Node.js test and Build
```
This doesn't seem to actually run...

sjrd replied:
It will run once the workflow hits `main`. It's always the same issue when we bootstrap a project.
I linked a green build of the same commit in my `sjrd/*` repo: https://github.com/sjrd/vite-plugin-scalajs/actions/runs/4376164277

gzm0 commented on test/plugin.test.ts:
```diff
@@ -0,0 +1,79 @@
+    await plugin.buildStart.call(fakePluginContext, {});
+
+    expect(normalizeSlashes(await plugin.resolveId.call(fakePluginContext, 'scalajs:main.js')))
+      .toContain('/testproject/target/scala-3.2.2/testproject-opt/main.js');
```
Have you considered checking for strict equality using `path.join`:

```ts
const expected = path.join(cwd, "target", "scala-3.2.2", "testproject-opt", "main.js")
```

sjrd replied:
The Vite documentation says that it internally normalizes paths to using slashes, but it's unclear when and how. It seems more consistent to me to write tests with `/` everywhere, to be closer to that idea.

gzm0 commented on test/plugin.test.ts:
```diff
@@ -0,0 +1,79 @@
+  return path === null ? null : path.replace(/\\/g, '/');
+}
+
+describe("scalaJSPlugin", () => {
```
Add a test case that sets `isDev` to true?

gzm0 commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+export default function scalaJSPlugin(options: ScalaJSPluginOptions = {}): VitePlugin {
+  const { projectID, cwd } = options;
+
+  let isDev = false; // when resolving the Vite config, this may turn into true
```
Consider being stricter here and type `isDef: boolean | undefined`. Then fail if `configResolved` didn't get called.

gzm0 commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+  const { projectID, cwd } = options;
+
+  let isDev = false; // when resolving the Vite config, this may turn into true
+  let scalaJSOutputDir: string;
```
Same here actually: Don't resolve to the empty string if you didn't get the callbacks you expected.

gzm0 reviewed (changes requested):
All in all LG. Just some nits and details.

sjrd commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+    },
+
+    // standard Rollup
+    resolveId(source, importer, options) {
```
No, we don't. If the default `resolveDynamicImport` fails to resolve something (across all plugins), then Vite falls back on `resolveId`. So for behavior that applies to both, we only need to implement `resolveId`.

sjrd commented on package.json:
```diff
@@ -0,0 +1,46 @@
+  },
+  "dependencies": {
+    "vite": "^4.1.4",
+    "vitest": "^0.29.2"
```
No, indeed. I moved it to `devDependencies`.

sjrd commented on tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+    "./*.ts"
+  ],
+  "exclude": [
+    "__test__/**.ts",
```
Cargo-cult copy-paste. :p I removed it.

sjrd commented on index.ts:
```diff
@@ -0,0 +1,72 @@
+
+    // standard Rollup
+    resolveId(source, importer, options) {
+      const parts = /^scalajs:(.*)$/g.exec(source);
```
I added a `uriPrefix` configuration for this.

sjrd commented:
Thanks for the review! I believe I have addressed all the comments.
New green build available at: https://github.com/sjrd/vite-plugin-scalajs/actions/runs/4391908549
